redis_session.c

```diff
@@ -183,6 +185,162 @@ redis_pool_get_sock(redis_pool *pool, const char *key TSRMLS_DC) {
+    smart_string_0(&lock_status->lock_key);
+
+    if (lock_expire > 0) {
+        cmd_len = REDIS_SPPRINTF(&cmd, "SET", "ssssd", lock_status->lock_key.c, lock_status->lock_key.len, lock_status->lock_secret.c, lock_status->lock_secret.len, "NX", 2, "PX", 2, lock_expire * 1000);
+    } else {
+        cmd_len = REDIS_SPPRINTF(&cmd, "SET", "sss", lock_status->lock_key.c, lock_status->lock_key.len, lock_status->lock_secret.c, lock_status->lock_secret.len, "NX", 2);
+    }
+
+    for (i_lock_retry = 0; !lock_status->is_locked && (max_lock_retries == -1 || i_lock_retry <= max_lock_retries); i_lock_retry++) {
+      if(!(redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC) < 0)
+          && ((response = redis_sock_read(redis_sock, &response_len TSRMLS_CC)) != NULL)
+          && response_len == 3
+          && strncmp(response, "+OK", 3) == 0) {
+            lock_status->is_locked = 1;
+      } else if (max_lock_retries == -1 || i_lock_retry < max_lock_retries) {
+          usleep(lock_wait_time);
+      }
+    }
+
+    if (response != NULL) {
+        efree(response);
+    }
+    efree(cmd);
+
+    if (lock_status->is_locked) {
+        return SUCCESS;
+    } else {
+        return FAILURE;
+    }
+}
+
+void refresh_lock_status(RedisSock *redis_sock, redis_session_lock_status *lock_status TSRMLS_DC)
+{
+    if (!lock_status->is_locked) return;
+    // If redis.session.lock_expire is not set => TTL=max_execution_time
+    // Therefore it is guaranteed that the current process is still holding the lock
+    if (lock_status->is_locked && INI_INT("redis.session.lock_expire") == 0) return;
+
+    char *cmd, *response;
+    int response_len, cmd_len;
+
+    cmd_len = REDIS_SPPRINTF(&cmd, "GET", "s", lock_status->lock_key.c, lock_status->lock_key.len);
+
+    redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
+    response = redis_sock_read(redis_sock, &response_len TSRMLS_CC);
+
+    if (response != NULL) {
+        lock_status->is_locked = (strcmp(response, lock_status->lock_secret.c) == 0);
+        efree(response);
+    } else {
+        lock_status->is_locked = 0;
+    }
+    efree(cmd);
+}
+
+int write_allowed(RedisSock *redis_sock, redis_session_lock_status *lock_status TSRMLS_DC)
+{
+    if (!INI_INT("redis.session.locking_enabled")) return 1;
+
+    refresh_lock_status(redis_sock, lock_status TSRMLS_CC);
+    return lock_status->is_locked;
+}
+
+void lock_release(RedisSock *redis_sock, redis_session_lock_status *lock_status TSRMLS_DC)
+{
+    if (lock_status->is_locked) {
+        char *cmd, *response;
+        int response_len, cmd_len;
+
+        upload_lock_release_script(redis_sock TSRMLS_CC);
+        cmd_len = REDIS_SPPRINTF(&cmd, "EVALSHA", "sdss", REDIS_G(lock_release_lua_script_hash), strlen(REDIS_G(lock_release_lua_script_hash)), 1, lock_status->lock_key.c, lock_status->lock_key.len, lock_status->lock_secret.c, lock_status->lock_secret.len);
+
+        redis_sock_write(redis_sock, cmd, cmd_len TSRMLS_CC);
```

Comment:
Don't ignore return value of `redis_sock_write`

Reply:
Will be fixed + additional error handling to check if just LUA script was not uploaded or an other error occured.